src/run/runner/helpers/isolation.rs

```diff
@@ -0,0 +1,63 @@
+use super::setup::run_with_sudo;
+
+pub struct CpuIsolation {
+    nproc: u32,
+}
+
+impl CpuIsolation {
+    pub fn new() -> Self {
+        Self {
+            nproc: num_cpus::get() as u32,
+        }
+    }
+
+    pub fn nproc(&self) -> u32 {
+        self.nproc
+    }
+
+    fn set_allowed_cpus(scope: &str, allowed_cpus: &[u32]) -> anyhow::Result<()> {
+        let cpus = allowed_cpus
+            .iter()
+            .map(|cpu| cpu.to_string())
+            .collect::<Vec<_>>()
+            .join(",");
+
+        run_with_sudo(&[
+            "systemctl",
+            "set-property",
+            scope,
+            &format!("AllowedCPUs={}", cpus),
+        ])?;
+
+        Ok(())
+    }
+
+    pub fn isolate_system(&self, allowed_cpus: &[u32]) -> anyhow::Result<()> {
+        Self::set_allowed_cpus("system.slice", allowed_cpus)?;
+        Self::set_allowed_cpus("init.scope", allowed_cpus)?;
+
+        Ok(())
+    }
+
+    pub fn isolate_user(&self, allowed_cpus: &[u32]) -> anyhow::Result<()> {
+        Self::set_allowed_cpus("user.slice", allowed_cpus)?;
+
+        Ok(())
+    }
+
+    pub fn reset_isolation(nproc: u32) -> anyhow::Result<()> {
```

Comment:
There should be no argument, and we should take nproc from self to make sure one can't mess up this call